llvm/lib/LTO/LTO.cpp

```diff
@@ -1645,19 +1648,48 @@ class WriteIndexesThinBackend : public ThinBackendProc {
       *LinkedObjectsFile << LinkedObjectsFilePath << '\n';
     }
 
-    if (auto E = emitFiles(ImportList, ModulePath, NewModulePath))
-      return E;
+    BackendThreadPool.async(
+        [this](const StringRef ModulePath,
+               const FunctionImporter::ImportMapTy &ImportList,
+               const std::string &OldPrefix, const std::string &NewPrefix) {
+          std::string NewModulePath =
+              getThinLTOOutputFile(ModulePath, OldPrefix, NewPrefix);
+          auto E = emitFiles(ImportList, ModulePath, NewModulePath);
+          if (E) {
+            std::unique_lock<std::mutex> L(ErrMu);
+            if (Err)
+              Err = joinErrors(std::move(*Err), std::move(E));
+            else
+              Err = std::move(E);
+            return;
+          }
+          if (OnWrite) {
```

Comment:
The InProcessThinBackend handles this by calling OnWrite after the async invocation. Can we use the same approach here?

Reply:
I tried that at first, but that would be a change in behavior and causes this test to fail here: https://github.com/llvm/llvm-project/blob/main/lld/test/ELF/lto/thinlto-emit-imports.ll#L12-L13

We should only call `OnWrite` if `emitFiles` succeeded, returning early if not.

Reply:
Do we get a different error? It would be good to avoid the locking.

Reply:
> Do we get a different error?

We did at the time, but I've changed it so we get the same error message, just wrapped in `'` and removed the locking.